### PR TITLE
Block state code generation

### PR DESCRIPTION
**Pull request by Trompettesib**

### CODE REVIEW

File: generators/final.hpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        case 7050:
+        case 7051:
+        case 7052:
+        case 7053:
+        case 7054:
+        case 7055:
+        case 7056:
+        case 7057:
+        case 7058:
+            return "minecraft::acacia_button";
```
Comment (STMiki) [inline generators/final.hpp:195]: `minecraft:acacio_button`

**Review by STMiki — COMMENTED** [review]
Add the blocks.json from mc to the gitignore.

**Comment by Trompettesib** [comment]
I think I have to concat all the toName function but I'm not sure about that

**Review by STMiki — DISMISSED** [review]
Looks good, but does a test compile ?

**Comment by Trompettesib** [comment]
> Looks good, but does a test compile ?

nop because of the toName functions

**Comment by Trompettesib** [comment]
> Add the blocks.json from mc to the gitignore.

Do I need to delete it on github ?

**Comment by Trompettesib** [comment]
Now it's compiling !

**Comment by STMiki** [comment]
> > Add the blocks.json from mc to the gitignore.
> 
> Do I need to delete it on github ?

I think so ? but this will be generated from the

**Comment by STMiki** [comment]
Oops

**Review by STMiki — COMMENTED** [review]
Mhh.. I don't know how to go about this.

Either let the generated.hpp file here,
or add it to the gitignore and generate it from the cmake.

But if we generate it, we need to download the .jar, extract the data and generate the .hpp

🤷

**Comment by Trompettesib** [comment]
I have a question.
In the function toName, do I have to return a std::pair<std::string, std::vector<std::pair<std::string, std::string>>> ?
Like, I'm only returning the name of the block for any protocol id it has. Maybe it should be better to return the name, and the properties if the block has some ? as in the fromNameToProtocolId function

**Comment by STMiki** [comment]
I think you should hide these in a block class

File: generators/protocol_block_ids.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+            data += "case " + str(state["id"]) + ":\n"
+            data += "return {\"" + self.name + "\", {"
+            if self.properties != []:
+                for prop in self.properties:
+                    data += "{\"" + prop + "\", \"" + state["properties"][prop] + "\"}, "
+                data = data[:-2]
+            data += "}};\n"
+        return data
+
+blocks = []
```
Comment (STMiki) [inline generators/protocol_block_ids.py:192]: Put theses in a function, because this script could be called from another one idk.

Comment (STMiki) [reply generators/protocol_block_ids.py:192]: call it from `if __name__ == "__main__":` thingy

**Review by STMiki — COMMENTED** [review]
👌

**Review by STMiki — APPROVED** [review]
👌

**Comment by huntears** [comment]
I also forgot but documenting the generator a little bit would be nice ;)

File: generators/protocol_block_ids.py [hunk 3]
```diff
@@ -264,25 +250,54 @@ def write_header_bis(blocks):
...
+
+# this is the main function
 def main():
-    blocks = load_json("blocks.json")
-    write_header(blocks)
-    write_header_bis(blocks)
+    options = {"input": "blocks.json", "output": "generated/blocks"}
+    options = parse_args(options)
+    blocks = load_json(options["input"])
+    os.makedirs('/'.join(options["output"].split('/')[:-1]), exist_ok=True)
```
Comment (STMiki) [inline generators/protocol_block_ids.py:298]: `os.path.dirname`

**Review by STMiki — APPROVED** [review]
Don't mind me, this is just me giving too much attention to minor detail... :upside_down_face:

File: generators/protocol_block_ids.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    parser.add_argument("-f", "--format", help="format the code", action=argparse.BooleanOptionalAction)
+    args = parser.parse_args()
+    if args.input != "blocks.json":
+        print("I hope you know what you are doing, cause for now the script only works with the blocks.json file")
+    options["input"] = args.input
+    options["output"] = args.output
+    formated = args.format
+    return options
+
+# this is the main function
```
Comment (huntears) [inline generators/protocol_block_ids.py:293]: Thank you captain obvious xd

**Review by huntears — CHANGES_REQUESTED** [review]
Where is the generated code ?

**Review by STMiki — APPROVED** [review]
👌

File: generators/protocol_block_ids.py (hunk 4 above)
Comment (Trompettesib) [reply generators/protocol_block_ids.py]: You're welcome ^^

**Comment by Trompettesib** [comment]
I almost did some tests but then I realize that I have already passed more than twice the time allowed for this task

**Comment by Trompettesib** [comment]
What append if the file blocks.json does not exist ?

**Review by STMiki — CHANGES_REQUESTED** [review]

File: cubic-server/Server.cpp [hunk 5]
```diff
@@ -145,3 +160,42 @@ void Server::_stop()
         delete command;
     LINFO("Server stopped");
 }
+
+void Server::_downloadFile(const std::string &url, const std::string &path)
+{
+    if (std::filesystem::exists(path)) {
+        LDEBUG("File " << path << " already exists. Skipping download");
```
Comment (STMiki) [inline cubic-server/Server.cpp:167]:
```suggestion
        LDEBUG("File " << path << " already exists. Skipping download");
        return;
```
and then remove the else 💢

File: cubic-server/protocol_id_converter/blockStates.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+        b.baseProtocolId = block.value()["states"][0]["id"];
+        b.maxProtocolId = block.value()["states"][block.value()["states"].size() - 1]["id"];
+        this->_blocks.push_back(b);
+    }
+}
+
+Blocks::BlockId Blocks::GlobalPalette::fromBlockToProtocolId(Blocks::Block block) const {
+    auto internalBlock = std::find_if(this->_blocks.begin(), this->_blocks.end(), [&block](Blocks::InternalBlock b) {
```
Comment (STMiki) [inline cubic-server/protocol_id_converter/blockStates.cpp:38]:
```suggestion
    auto internalBlock = std::find_if(this->_blocks.begin(), this->_blocks.end(), [&block](const Blocks::InternalBlock &b) {
```

File: cubic-server/protocol_id_converter/blockStates.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    if (block.properties.size() != internalBlock->properties.size()) {
+        LERROR("Block properties size mismatch with block " << block.name);
+        return 0;
+    }
+    if (block.properties.size() == 0 && internalBlock->properties.size() == 0)
+        return internalBlock->baseProtocolId;
+
+    BlockId id = internalBlock->baseProtocolId;
+    for (auto property : block.properties) {
+        auto internalProperty = std::find_if(internalBlock->properties.begin(), internalBlock->properties.end(), [&property](Blocks::InternalProperty p) {
```
Comment (STMiki) [inline cubic-server/protocol_id_converter/blockStates.cpp:54]:
```suggestion
        auto internalProperty = std::find_if(internalBlock->properties.begin(), internalBlock->properties.end(), [&property](const Blocks::InternalProperty &p) {
```

File: cubic-server/protocol_id_converter/blockStates.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            return 0;
+        }
+        id += internalProperty->baseWeight * (value - internalProperty->values.begin());
+    }
+    return id;
+}
+
+Blocks::Block Blocks::GlobalPalette::fromProtocolIdToBlock(Blocks::BlockId id) const {
+    for (auto b : this->_blocks) {
+        if (id >= b.baseProtocolId && id <= b.maxProtocolId) {
```
Comment (STMiki) [inline cubic-server/protocol_id_converter/blockStates.cpp:73]:
```suggestion
        if (id < b.baseProtocolId || id > b.maxProtocolId)
                continue;
```

File: cubic-server/protocol_id_converter/blockStates.hpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    class GlobalPalette {
+        public:
+            GlobalPalette() {};
+            ~GlobalPalette() {};
+            /**
+             * @brief Convert a block to a protocol id
+             * @param block The block to convert
+             * @return The protocol id of the block
+             */
+            BlockId fromBlockToProtocolId(Block block) const;
```
Comment (STMiki) [inline cubic-server/protocol_id_converter/blockStates.hpp:50]:
```suggestion
            BlockId fromBlockToProtocolId(const Block &block) const;
```

**Comment by STMiki** [comment]
It fails the test, and you could have a base class with inerhitance from both items and block, I think, they are really similar

File: .devcontainer/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,27 @@
+FROM mcr.microsoft.com/devcontainers/cpp:0-ubuntu-22.04
```
[the remark this answers is not in the document]
Comment (Trompettesib) [reply .devcontainer/Dockerfile:1]: It's if you want to use github codespace. It's the default image.

File: .github/workflows/publish.yml [hunk 11]
```diff
@@ -27,7 +27,7 @@ jobs:
       if: ${{ !contains(github.repository, 'self-hosted') }}
       run: |
         sudo apt-get update
-        sudo apt-get install -y gcc-12 g++-12 cmake pkg-config gtkmm-3.0
+        sudo apt-get install -y gcc-12 g++-12 cmake pkg-config gtkmm-3.0 libcurl4-openssl-dev
```
[the remark this answers is not in the document]
Comment (Trompettesib) [reply .github/workflows/publish.yml:30]: Here there isn't clang ^^

**Comment by Trompettesib** [comment]
> Do you still need `generators/requirements.txt` ?

yes because I am using some of the libraries in the python file.
Maybe in the future, we will rethink about the code generation and reuse the python file

File: .devcontainer/Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (STMiki) [reply .devcontainer/Dockerfile:1]: I mean the file 😆

File: cubic-server/protocol_id_converter/blockStates.cpp (hunk 8 above)
Comment (Trompettesib) [reply cubic-server/protocol_id_converter/blockStates.cpp:73]: Isn't this dangerous to let a return keyword in the for loop without having the if scope ?

Comment (huntears) [reply cubic-server/protocol_id_converter/blockStates.cpp:73]: Nop

File: cubic-server/Server.cpp (hunk 5 above)
Comment (huntears) [reply cubic-server/Server.cpp:167]: I don't think that's needed to be fair

File: .github/workflows/publish.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (STMiki) [reply .github/workflows/publish.yml:30]: I need more sleep...

File: cubic-server/protocol_id_converter/blockStates.hpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "nlohmann/json.hpp"
+
+namespace Blocks {
+    // Using int32_t because it's the same size as a varint
+    typedef int32_t BlockId;
```
Comment (STMiki) [inline cubic-server/protocol_id_converter/blockStates.hpp:9]: Either use this everywhere and remove `GlobalBlockId` in `types.hpp` or use `GlobalBlockId` here

Comment (Trompettesib) [reply cubic-server/protocol_id_converter/blockStates.hpp:9]: but GlobalBlockId is a uint16_t...
```cpp
typedef uint16_t GlobalBlockId;
```

Comment (Trompettesib) [reply cubic-server/protocol_id_converter/blockStates.hpp:9]: How do I need to change it ?

Comment (STMiki) [reply cubic-server/protocol_id_converter/blockStates.hpp:9]: Wut ?

Comment (STMiki) [reply cubic-server/protocol_id_converter/blockStates.hpp:9]: wtf, I use this in varInt, this should be int32_t wtf

Comment (Trompettesib) [reply cubic-server/protocol_id_converter/blockStates.hpp:9]: https://github.com/CubicMC/cubic-server/blob/e92ff032f8bb4aa6d99d9cb207a99e04d5ba5d89/cubic-server/types.hpp#L15

Comment (Trompettesib) [reply cubic-server/protocol_id_converter/blockStates.hpp:9]: Do you want me to change that ?

Comment (STMiki) [reply cubic-server/protocol_id_converter/blockStates.hpp:9]: Ok no, I add them to the palette and convert them to an `int32_t` so you can just transform them directly

**Review by huntears — APPROVED** [review]

**Review by STMiki — APPROVED** [review]
👌

**Pull request merged**